File: models/smartspaces-interaction/src/main/java/io/smartspaces/sandbox/interaction/processing/sensor/StandardBleProximitySensorValueProcessor.java
/*
 * Copyright (C) 2016 Keith M. Hughes
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package io.smartspaces.sandbox.interaction.processing.sensor;

import java.util.HashMap;
import java.util.Map;

import io.smartspaces.event.trigger.SimpleHysteresisThresholdValueTrigger;
import io.smartspaces.event.trigger.Trigger;
import io.smartspaces.event.trigger.TriggerEventType;
import io.smartspaces.event.trigger.TriggerListener;
import io.smartspaces.event.trigger.TriggerState;
import io.smartspaces.sandbox.interaction.entity.MarkableEntityDescription;
import io.smartspaces.sandbox.interaction.entity.MarkerEntityDescription;
import io.smartspaces.sandbox.interaction.entity.SensorEntityDescription;
import io.smartspaces.sandbox.interaction.entity.model.PersonSensedEntityModel;
import io.smartspaces.sandbox.interaction.entity.model.PhysicalSpaceSensedEntityModel;
import io.smartspaces.sandbox.interaction.entity.model.SensedEntityModel;
import io.smartspaces.sandbox.interaction.entity.model.updater.PersonPhysicalSpaceModelUpdater;
import io.smartspaces.sandbox.interaction.entity.sensor.StandardSensorData;
import io.smartspaces.util.data.dynamic.DynamicObject;
import io.smartspaces.util.data.dynamic.StandardDynamicObjectNavigator;

/**
 * The standard processor for BLE proximity data.
 * 
 * @author Keith M. Hughes
 */
public class StandardBleProximitySensorValueProcessor implements SensorValueProcessor {

  /**
   * The map from the BLE IDs to the trigger for that ID.
   */
  private Map<String, SimpleHysteresisThresholdValueTrigger> userTriggers = new HashMap<>();

  /**
   * The map from the triggers to the models to be updated by that trigger.
   */
  private Map<SimpleHysteresisThresholdValueTrigger, PersonPhysicalSpaceModelUpdater> userTriggerToUpdaters =
      new HashMap<>();

  /**
   * The listener for trigger events being shared across all triggers.
   */
  private TriggerListener triggerListener = new TriggerListener() {
    @Override
    public void onTrigger(Trigger trigger, TriggerState state, TriggerEventType type) {
      handleTrigger(trigger, state, type);
    }
  };

  /**
   * support for working with BLE proximity devices.
   */
  private StandardBleProximitySupport bleProximitySupport = new StandardBleProximitySupport();

  @Override
  public String getSensorValueType() {
    return StandardSensorData.SENSOR_TYPE_PROXIMITY_BLE;
  }

  @Override
  public void processData(long timestamp, SensorEntityDescription sensor,
      SensedEntityModel sensedEntityModel, SensorValueProcessorContext processorContext,
      DynamicObject data) {
    String markerId = "ble" + ":" + data.getRequiredString("id");
    double rssi = data.getDouble("rssi");

    SimpleHysteresisThresholdValueTrigger userTrigger =
        getTrigger(markerId, sensor, sensedEntityModel, processorContext);
    userTrigger.update(rssi);

    MarkableEntityDescription markedEntity = processorContext.getSensedEntityModelCollection()
        .getSensorRegistry().getMarkableEntityByMarkerId(markerId);
    processorContext.getLog().formatInfo("Detected ID %s,  RSSI= %f, %s\n", markerId, rssi,
        markedEntity);
  }

  /**
   * Get the trigger for a given marker ID.
   * 
   * <p>
   * Creates the trigger if it didn't exist.
   * 
   * @param markerId
   *          the marker ID for the trigger
   * @param sensedEntityModel
   *          the sensed entity model that is associated with the sensor
   * @param processorContext
   *          the context for processor handling
   * 
   * @return the trigger for the marker
   */
  private SimpleHysteresisThresholdValueTrigger getTrigger(String markerId,
      SensorEntityDescription sensor, SensedEntityModel sensedEntityModel,
      SensorValueProcessorContext processorContext) {
    SimpleHysteresisThresholdValueTrigger userTrigger = userTriggers.get(markerId);
    if (userTrigger == null) {
      userTrigger = new SimpleHysteresisThresholdValueTrigger();

      MarkerEntityDescription markerEntity = processorContext.getSensedEntityModelCollection()
          .getSensorRegistry().getMarkerEntityByMarkerId(markerId);

      Map<String, Object> configData = processorContext.getSensedEntityModelCollection()
          .getSensorRegistry().getConfigurationData(markerEntity.getId());
      bleProximitySupport.configureTrigger(userTrigger, configData, sensor, processorContext);

      userTrigger.addListener(triggerListener);
      userTriggers.put(markerId, userTrigger);

      PersonSensedEntityModel person =
          processorContext.getSensedEntityModelCollection().getMarkedSensedEntityModel(markerId);
      PersonPhysicalSpaceModelUpdater modelUpdater = new PersonPhysicalSpaceModelUpdater(
          (PhysicalSpaceSensedEntityModel) sensedEntityModel, person);
      userTriggerToUpdaters.put(userTrigger, modelUpdater);
    }

    return userTrigger;
  }

  /**
   * Handle a trigger change,
   * 
   * @param trigger
   *          the trigger that changed
   * @param state
   *          the new state of the trigger
   * @param type
   *          the type of the state change
   */
  private void handleTrigger(Trigger trigger, TriggerState state, TriggerEventType type) {
    PersonPhysicalSpaceModelUpdater modelUpdater = userTriggerToUpdaters.get(trigger);
    if (type == TriggerEventType.RISING) {
      modelUpdater.enterSpace();
    } else {
      modelUpdater.exitSpace();
    }
  }
}

File: models/smartspaces-interaction/src/main/scala/io/smartspaces/sandbox/sensor/processing/StandardBleProximitySensorValueProcessor.scala
/*
 * Copyright (C) 2016 Keith M. Hughes
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package io.smartspaces.sandbox.sensor.processing

import io.smartspaces.event.trigger.SimpleHysteresisThresholdValueTriggerWithData
import io.smartspaces.event.trigger.TriggerWithData
import io.smartspaces.event.trigger.TriggerEventType
import io.smartspaces.event.trigger.TriggerWithDataListener
import io.smartspaces.event.trigger.TriggerState
import io.smartspaces.util.data.dynamic.DynamicObject
import io.smartspaces.sandbox.sensor.entity.model.updater.SimplePersonPhysicalSpaceModelUpdater
import io.smartspaces.sandbox.sensor.entity.model.SensedEntityModel
import io.smartspaces.sandbox.sensor.entity.model.PersonSensedEntityModel
import io.smartspaces.sandbox.sensor.entity.model.SensorEntityModel
import io.smartspaces.sandbox.sensor.entity.model.PhysicalSpaceSensedEntityModel
import io.smartspaces.sandbox.sensor.StandardSensorData

import scala.collection.mutable._

/**
 * The standard processor for BLE proximity data.
 *
 * @author Keith M. Hughes
 */
class StandardBleProximitySensorValueProcessor extends SensorValueProcessor {

  /**
   * The map from the BLE IDs to the trigger for that ID.
   */
  private val userTriggers: Map[String, SimpleHysteresisThresholdValueTriggerWithData[Long]] = new HashMap

  /**
   * The map from the triggers to the models to be updated by that trigger.
   */
  private val userTriggerToUpdaters: Map[SimpleHysteresisThresholdValueTriggerWithData[Long], SimplePersonPhysicalSpaceModelUpdater] =
    new HashMap

  /**
   * The listener for trigger events being shared across all triggers.
   */
  private val triggerListener = new TriggerWithDataListener[Long]() {
    override def onTrigger(trigger: TriggerWithData[Long], state: TriggerState, eventType: TriggerEventType): Unit = {
      handleTrigger(trigger, state, eventType);
    }
  };

  /**
   * support for working with BLE proximity devices.
   */
  private var bleProximitySupport: StandardBleProximitySupport = new StandardBleProximitySupport()

  val sensorValueType = StandardSensorData.SENSOR_TYPE_PROXIMITY_BLE

  override def processData(timestamp: Long, sensor: SensorEntityModel,
    sensedEntityModel: SensedEntityModel, processorContext: SensorValueProcessorContext,
    data: DynamicObject) {
    val markerId = "ble" + ":" + data.getRequiredString("id")
    val rssi = data.getDouble("rssi")

    val userTrigger = getTrigger(markerId, sensor, sensedEntityModel, processorContext)
    userTrigger.update(rssi, timestamp)

    val markedEntity = processorContext.completeSensedEntityModel.
      sensorRegistry.getMarkableEntityByMarkerId(markerId)
    processorContext.log.formatInfo("Detected ID %s,  RSSI= %f, %s\n", markerId, rssi,
      markedEntity);
  }

  /**
   * Get the trigger for a given marker ID.
   *
   * <p>
   * Creates the trigger if it didn't exist.
   *
   * @param markerId
   *          the marker ID for the trigger
   * @param sensedEntityModel
   *          the sensed entity model that is associated with the sensor
   * @param processorContext
   *          the context for processor handling
   *
   * @return the trigger for the marker
   */
  private def getTrigger(markerId: String,
    sensor: SensorEntityModel, sensedEntityModel: SensedEntityModel,
    processorContext: SensorValueProcessorContext): SimpleHysteresisThresholdValueTriggerWithData[Long] = {
    val userTrigger = userTriggers.get(markerId)
    if (userTrigger.isEmpty) {
      val newUserTrigger = new SimpleHysteresisThresholdValueTriggerWithData[Long](0)

      val markerEntity = processorContext.completeSensedEntityModel.
        sensorRegistry.getMarkerEntityByMarkerId(markerId)

      val configData: scala.collection.immutable.Map[String, Object] = processorContext.completeSensedEntityModel.
        sensorRegistry.getConfigurationData(markerEntity.get.id)
      bleProximitySupport.configureTrigger(newUserTrigger, configData, sensor, processorContext);

      newUserTrigger.addListener(triggerListener)
      userTriggers.put(markerId, newUserTrigger)

      val person =
        processorContext.completeSensedEntityModel.getMarkedSensedEntityModel(markerId).asInstanceOf[PersonSensedEntityModel]
      val modelUpdater = new SimplePersonPhysicalSpaceModelUpdater(
        sensedEntityModel.asInstanceOf[PhysicalSpaceSensedEntityModel], person)
      userTriggerToUpdaters.put(newUserTrigger, modelUpdater)

      newUserTrigger
    } else {
      userTrigger.get
    }
  }

  /**
   * Handle a trigger change,
   *
   * @param trigger
   *          the trigger that changed
   * @param state
   *          the new state of the trigger
   * @param triggerType
   *          the type of the state change
   */
  private def handleTrigger(trigger: TriggerWithData[Long], state: TriggerState, eventType: TriggerEventType): Unit = {
    val t = trigger.asInstanceOf[SimpleHysteresisThresholdValueTriggerWithData[Long]]
    val modelUpdater = userTriggerToUpdaters.get(t).get
    if (eventType == TriggerEventType.RISING) {
      modelUpdater.enterSpace(t.getData);
    } else {
      modelUpdater.exitSpace(t.getData);
    }
  }
}
